### PR TITLE
feat(dashboard): Nv 4866 Enable override custom controls and implement custom controls for Delay & Digest

### PR DESCRIPTION
**Pull request by BiswaViraj**
### Screenshots

https://github.com/user-attachments/assets/f8ae4ef4-2529-470b-a150-fb7a5a4c9da0

<details>
<summary><strong>Expand for optional sections</strong></summary>

### Special notes for your reviewer

</details>

### CODE REVIEW

File: apps/dashboard/src/components/workflow-editor/steps/configure-step-form.tsx [hunk 1]
```diff
@@ -191,163 +193,165 @@ export const ConfigureStepForm = (props: ConfigureStepFormProps) => {
...
-              workflowSlug: workflow.slug,
-            })}
-            className="ml-auto flex items-center"
-          >
-            <Button variant="link" size="icon" className="size-4" type="button">
-              <RiCloseFill />
-            </Button>
-          </Link>
-        </SidebarHeader>
+      <AnimatePresence>
```
Comment (BiswaViraj) [inline apps/dashboard/src/components/workflow-editor/steps/configure-step-form.tsx:196]: This fixes the greyed out screen which sometimes happen while switching steps

File: apps/dashboard/src/components/workflow-editor/steps/configure-step-form.tsx [hunk 2]
```diff
@@ -191,163 +193,165 @@ export const ConfigureStepForm = (props: ConfigureStepFormProps) => {
...
+          {!isSupportedStep && (
+            <>
+              <SidebarContent>
+                <SdkBanner />
+              </SidebarContent>
+            </>
+          )}
+
+          {!isReadOnly && (
+            <>
```
Comment (SokratisVidros) [inline apps/dashboard/src/components/workflow-editor/steps/configure-step-form.tsx]: Please remove the extra fragment

File: apps/dashboard/src/components/workflow-editor/steps/configure-step-form.tsx [hunk 3]
```diff
@@ -191,163 +193,165 @@ export const ConfigureStepForm = (props: ConfigureStepFormProps) => {
...
-        <Separator />
+          <Separator />
+
+          <Form {...form}>
+            <form onBlur={onBlur}>
+              <SaveFormContext.Provider value={value}>
+                <SidebarContent>
+                  <FormField
+                    control={form.control}
+                    name={'name'}
```
Comment (SokratisVidros) [inline apps/dashboard/src/components/workflow-editor/steps/configure-step-form.tsx]:
```suggestion
                    name='name'
```

File: apps/dashboard/src/components/workflow-editor/steps/digest/digest-custom-control-values.tsx [hunk 4]
```diff
@@ -0,0 +1,39 @@
+import { Button } from '@/components/primitives/button';
+import { useWorkflow } from '@/components/workflow-editor/workflow-provider';
+import { RiCloseLine, RiEdit2Line } from 'react-icons/ri';
+import { useNavigate } from 'react-router-dom';
+import { CustomStepControls } from '../controls/custom-step-controls';
+
+export const DigestCustomControlValues = () => {
```
Comment (SokratisVidros) [inline apps/dashboard/src/components/workflow-editor/steps/digest/digest-custom-control-values.tsx]: Can we reuse this component across all steps?

File: apps/dashboard/src/hooks/use-form-autosave.ts [hunk 5]
```diff
@@ -20,13 +20,16 @@ export function useFormAutosave<U extends Record<string, unknown>, T extends Fie
   const formRef = useDataRef(propsForm);
 
   const onSave = useCallback(
-    async (data: T) => {
+    async (data: T, options?: { forceSubmit?: boolean }) => {
```
Comment (SokratisVidros) [inline apps/dashboard/src/hooks/use-form-autosave.ts:23]: @BiswaViraj what is the use case for the forceSubmit option?

**Review by SokratisVidros — CHANGES_REQUESTED** [review]

File: apps/dashboard/src/hooks/use-form-autosave.ts (hunk 5 above)
Comment (BiswaViraj) [reply apps/dashboard/src/hooks/use-form-autosave.ts:23]: When the user toggle off the override, we reset the form with the custom control default values and discard their modified values.
Now resetting the form doesn't makes the form dirty, and we have a condition which checks if the form is dirty before submitting which is what we want for rest of the cases.
but in this scenario we need to bypass the isDirty check and submit the defaultValues 
@SokratisVidros

File: apps/dashboard/src/components/workflow-editor/steps/digest/digest-custom-control-values.tsx (hunk 4 above)
Comment (BiswaViraj) [reply apps/dashboard/src/components/workflow-editor/steps/digest/digest-custom-control-values.tsx]: No, but can be with delay and digest, updating it to use within delay and digest

File: apps/dashboard/src/components/workflow-editor/steps/controls/custom-step-controls.tsx [hunk 6]
```diff
@@ -1,52 +1,208 @@
...
+const CONTROLS_DOCS_LINK = 'https://docs.novu.co/concepts/controls';
+
 export const CustomStepControls = (props: CustomStepControlsProps) => {
   const { className, dataSchema, origin, ...rest } = props;
   const [isEditorOpen, setIsEditorOpen] = useState(true);
+  const [isRestoreDefaultModalOpen, setIsRestoreDefaultModalOpen] = useState(false);
+  const { step } = useWorkflow();
+  const { reset } = useFormContext();
+  const { saveForm } = useSaveForm();
+  const overrideForm = useForm({ defaultValues: { override: Object.keys(step?.controls.values ?? {}).length > 0 } });
```
Comment (LetItRock) [inline apps/dashboard/src/components/workflow-editor/steps/controls/custom-step-controls.tsx]: why do we need a nested form here? it seems that we can just have a state

File: apps/dashboard/src/components/workflow-editor/steps/controls/custom-step-controls.tsx [hunk 7]
```diff
@@ -1,52 +1,208 @@
...
+      <SidebarContent size="md">
+        <Collapsible
+          open={isEditorOpen}
+          onOpenChange={setIsEditorOpen}
+          className={cn(
+            'bg-neutral-alpha-50 border-neutral-alpha-200 flex w-full flex-col gap-2 rounded-lg border p-2 text-sm',
+            className
+          )}
+          {...rest}
+        >
```
Comment (LetItRock) [inline apps/dashboard/src/components/workflow-editor/steps/controls/custom-step-controls.tsx]: why we are not using accordion primitive that has a nice animation?

File: apps/dashboard/src/components/workflow-editor/steps/controls/custom-step-controls.tsx [hunk 8]
```diff
@@ -1,52 +1,208 @@
...
+    <SidebarContent size="md">
+      <ConfirmationModal
+        open={isRestoreDefaultModalOpen}
+        onOpenChange={setIsRestoreDefaultModalOpen}
+        onConfirm={async () => {
+          const defaultValues = buildDefaultValuesOfDataSchema(step?.controls.dataSchema ?? {});
+          reset(defaultValues);
+          saveForm(true);
+          setIsRestoreDefaultModalOpen(false);
+          overrideForm.setValue('override', false);
```
Comment (LetItRock) [inline apps/dashboard/src/components/workflow-editor/steps/controls/custom-step-controls.tsx]: have you tried to use the `reset` function options? the `saveForm(true)` seems a bit of a workaround

**Review by LetItRock — CHANGES_REQUESTED** [review]
Screenshot 2024-12-16 at 10 51 08

File: apps/dashboard/src/components/workflow-editor/steps/controls/custom-step-controls.tsx (hunk 7 above)
Comment (BiswaViraj) [reply apps/dashboard/src/components/workflow-editor/steps/controls/custom-step-controls.tsx]: Didnt knew we had that, will update it to use the Accordion

File: apps/dashboard/src/components/workflow-editor/steps/controls/custom-step-controls.tsx (hunk 8 above)
Comment (BiswaViraj) [reply apps/dashboard/src/components/workflow-editor/steps/controls/custom-step-controls.tsx]: Yes, none of those options worked

**Comment by BiswaViraj** [comment]
@LetItRock The screenshot you added about the custom controls being visible is fixed in this PR https://github.com/novuhq/novu/pull/7302

File: apps/dashboard/src/hooks/use-form-autosave.ts (hunk 5 above)
Comment (SokratisVidros) [reply apps/dashboard/src/hooks/use-form-autosave.ts:23]: Thanks for the explanation. That's surprising for RHF. But anyway, in that case, maybe naming it to ignoreIsDirty would be easier to reason about.

File: apps/dashboard/src/components/workflow-editor/steps/controls/custom-step-controls.tsx (hunk 8 above)
Comment (LetItRock) [reply apps/dashboard/src/components/workflow-editor/steps/controls/custom-step-controls.tsx]: The form consists of:
```
{
name,
stepId,
controlValues: { ... }
}
```
It seems that you are not resetting it with a full-form object. If that won't work that you can try just to set the value with `setValue('controlValues', {})`.

Comment (BiswaViraj) [reply apps/dashboard/src/components/workflow-editor/steps/controls/custom-step-controls.tsx]: No actually, the form only contains the controlValues in a flat structure like
```
body, subject, customField, etc
```

File: apps/dashboard/src/components/workflow-editor/steps/configure-step-template-form.tsx [hunk 9]
```diff
@@ -18,15 +19,16 @@ import { Form } from '@/components/primitives/form/form';
...
   [StepTypeEnum.CHAT]: OtherStepTabs,
   [StepTypeEnum.IN_APP]: InAppTabs,
   [StepTypeEnum.SMS]: OtherStepTabs,
   [StepTypeEnum.PUSH]: OtherStepTabs,
-  [StepTypeEnum.DIGEST]: () => null,
-  [StepTypeEnum.DELAY]: () => null,
+  [StepTypeEnum.DIGEST]: CommonCustomControlValues,
+  [StepTypeEnum.DELAY]: CommonCustomControlValues,
   [StepTypeEnum.TRIGGER]: () => null,
   [StepTypeEnum.CUSTOM]: () => null,
```
Comment (LetItRock) [inline apps/dashboard/src/components/workflow-editor/steps/configure-step-template-form.tsx:33]: @BiswaViraj should we do the same for the `CUSTOM` step?

**Review by LetItRock — APPROVED** [review]

File: apps/dashboard/src/components/workflow-editor/steps/configure-step-template-form.tsx (hunk 9 above)
Comment (BiswaViraj) [reply apps/dashboard/src/components/workflow-editor/steps/configure-step-template-form.tsx:33]: I will need to check the whole flow for the custom step.
will do it in a separate PR

**Pull request merged**